### ISSUE TITLE
Suggestion: add support for http.cookieFile

### ISSUE THREAD

**Issue by tracervision** [libgit2/libgit2#3397]
It would be awesome if libgit2 had support for the http.cookieFile configuration. Tools like Xcode Bots use libgit2 and cookie support would be a big help in an environment where they are used for non-interactive repository access.

http://git-scm.com/docs/git-config

**Comment by carlosmn** [comment]
This is an area in git which is trying to expose libcurl functionality around HTTP, which we do not use so we can't just hand off a file.

What is it that you need this file for? Is authentication done via a cookie? How does it fail if you don't provide it? If you need to set a header field with a cookie, that's something I'd been thinking of adding, but wasn't sure how useful it would be.

**Comment by dborowitz** [comment]
Owner of the servers @tracervision is talking to speaking.

> This is an area in git which is trying to expose libcurl functionality around HTTP, which we do not use so we can't just hand off a file.

I understand that supporting this this is going to be more work than just setting a curl option. But I'm not sure I buy that the answer is to provide some alternative libgit2-specific configuration method--this is a documented config option in git-config(1), so if libgit2 wants to be compatible with git, it should use it. Surely there are other config options that libgit2 respects; where do you draw the line?

> Is authentication done via a cookie?

Yes.

> How does it fail if you don't provide it?

The HTTP request fails and you can't access the repository. Let's say it gives HTTP 403.

> If you need to set a header field with a cookie, that's something I'd been thinking of adding, but wasn't sure how useful it would be.

Not useful. We tell our users to set up http.cookiefile in ~/.gitconfig once and expect it to work everywhere. Even if we wanted to write the shim to read from cookies.txt, we don't own Xcode and can't modify its usage of libgit2 to make extra calls to set headers.

**Comment by ethomson** [comment]
Thanks for the additional background.

> so if libgit2 wants to be compatible with git, it should use it.

Well, that's not *quite* so straightforward, as you point out, as there *is* a line.

libgit2 doesn't (strictly speaking) want to be compatible with git.  We want to enable clients to be compatible with git.  There are a lot of configuration options that only really make sense in the context of the client itself and there are a lot of configuration options that are very leaky abstractions and don't really make sense for us.  This is especially true of many of the `http.*` configuration options.

`http.cookieFile` seems like a reasonable addition to me.

Now, having said that:  please file a radar bug about this.  Your bug is *with Apple's Xcode bot*.  Yes, they use libgit2 and yes, maybe if we do some work we could maybe enable this to work for you *maybe* without them doing any work, but that's quite a generous assumption.  Fundamentally, neither you nor I knowing how they're actually driving libgit2 today, let alone tomorrow.

I'm not trying to be unhelpful here; they could very easily be using their own network stack and any conversation we're having about this is moot.

If you're saying "wow, that's farfetched, nobody would do that", then I would point out that:

1. Every Microsoft tool that uses libgit2 does this so that we can easily handle authentication to Microsoft Accounts.
2. Apple does bizarre things with their version of git like hardcoding a credential helper for you that you can't disable.

So please file a radar bug independently of the question of what libgit2 should do.

**Comment by dborowitz** [comment]
You make some good points Ed. I suppose if Apple provided a way to configure the cookie file from the Xcode UI or something, that wouldn't quite meet the ideal of "configure once and forget it", but it would probably be ok.

I think we're in agreement that some of the http.* options are leaky (e.g. all the SSL crap) and some (e.g. http.cookiefile) less so. No need to draw a bright line today.

**Comment by ethomson** [comment]
Agreed.  Sorry to get all philosophical - I also don't want to draw a line today.  I'm not really sure I even know where it would be - I fear that this is one of those "we know it when we see it" sorts of things.  :)

**Comment by tracervision** [comment]
Right on. I agree that this issue is actually one for Apple. I'll file a radar and follow up with them directly. Thanks to everybody in the discussion.
